R/nimble_lm.R

```diff
@@ -5,40 +5,152 @@ library(assertthat) ## to package dependencies
 
 ## Helper Functions ##
 
-## This Function Makes the Name someFactor.effects
+## These Functions Make the Name someFactor.X
 make.effect.name <- function(term) {
-  paste0(term, '.effect')
+  as.name(paste0(term, '.effect'))
 }
 
+make.mean.name <- function(term) {
+  as.name(paste0(term, '.mu'))
+}
+
+make.precision.name <- function(term) {
+  as.name(paste0(term, '.tau'))
+}
+
+make.sigma.name <- function(term) {
+  as.name(paste0(term, '.sigma2'))
+}
+
+
+
 #Make Factor Variable
 makeFac <- function(x,char.only=FALSE) {
   if (!is.factor(x) && (!char.only || is.character(x))) factor(x) else x
 }
 
 
+#Extracts formula From Environment
+RHSForm <- function(form,as.form=FALSE) {
+  rhsf <- form[[length(form)]]
+  if (as.form) reformulate(deparse(rhsf)) else rhsf
+}
+
+
+#Functions for Random Effects
+
+#Replaces '|' with '+'
+
+subbars <- function(term)
+{
+  if (is.name(term) || !is.language(term)) return(term)
+  if (length(term) == 2) {
+    term[[2]] <- subbars(term[[2]])
+    return(term)
+  }
+  stopifnot(length(term) >= 3)
+  if (is.call(term) && term[[1]] == as.name('|'))
+    term[[1]] <- as.name('+')
+  for (j in 2:length(term)) term[[j]] <- subbars(term[[j]])
+  term
+}
+
+#Find Random Effect Terms from Formula object
+fb <- function(term)
+{
+  if (is.name(term) || !is.language(term)) return(NULL)
+  if (term[[1]] == as.name("(")) return(fb(term[[2]]))
```

Comment:
I've found that comparisons of pieces of R expressions can be error-prone.  If term[[1]] itself is an expression with length > 1, then it may not behave will with `==`.  You could wait and see, but if you get problems you can use either `identical(term[[1]], as.name("(")` or `deparse(term[[1]]) == "("`.